Add playerctl module

jdholtz opened the pull request:
Fixes #2205.

This PR adds support for the playerctl module. It uses [Playerctl's Python binding library](https://lazka.github.io/pgi-docs/Playerctl-2.0/) to monitor and extract information from players. Most of the configuration options are the same as the `cmus` module. There are three more added: `button_shift`, `button_unshift`, and `ignored_players`.

An important thing to note: If any text sent to i3bar contains an `&`, the text will not render. The `&` is common in videos and therefore creates a problem. I also see this issue in the `cmus` module, so it isn't directly an issue with this module, but should be solved to provide users with a better experience.

There are two ways I see solving this: 
1. Fix it in the playerctl module by doing some sort of `str.replace("&", "and")`
2. Fix it in py3status, which would also fix this issue for other modules

I also have a few more configuration options in mind but would like to get others' opinions:
1. A `max_width` option to render the text with an `...` if it is too long. There are a lot of videos that span half the i3bar. However, I am not sure if this can be handled in Py3status's formatter or not.
2. A `player_priority` option to initially select the active player. Currently, it selects the first one that playerctl selects. I don't think this option is necessary as I only see it being a factor during the post_config_hook (mostly only run when the user starts i3). After the initial start, active players will be cycled through by the most recent player to change status, metadata, or be opened. Additionally, configuration options are provided for the user to manually switch the active player (`button_shift` and `button_unshift`)

This might remove the need for the `mpris` module, but more information about a player is provided in that module so that be good to still keep. Additionally, the `player_control` module probably doesn't need to be kept as this module controls the same players.

Any feedback is appreciated. I will be willing to maintain this module (fixing bugs, adding new features, refactoring, etc.).

lasers commented:
>Fix it in py3status, which would also fix this issue for other modules

Does `{placeholder:escape}` works?

>A max_width option to render the text with an ... if it is too long. There are a lot of videos that span half the i3bar. However, I am not sure if this can be handled in Py3status's formatter or not.

I think it was brought up in the past and possibly coded, but not implemented.

>A player_priority option to initially select the active player. Currently, it selects the first one that playerctl selects. I don't think this option is necessary as...

Actually, I wanted to treat this as a loop of players, so all players that comes and go would appear on the bar along with its own buttons.

* See `nvidia_smi` for an example.. as it's done in a loop for those with two or more GPUs. The other example is `hddtemp` and `bluetooth` since there can be many hdds and connected devices.

If this provides to be a good replacement for mpris, then yes, we could revisit removing mpris in few years. Playerctl module would need to cover everything and be easily better than mpris out of the box.

jdholtz commented:
> Does {placeholder:escape} works?

Yes, that works perfectly!

> I think it [max_width] was brought up in the past and possibly coded, but not implemented.

If desired, I can make a PR implementing this config option for every module. It would be similar to the `min_width` option and show an `...` if it cuts off extra output. It could be helpful for other modules, especially music player modules.

> Actually, I wanted to treat this as a loop of players, so all players that comes and go would appear on the bar along with its own buttons.

Sounds good. I'll make the change. How would buttons be done? None of the modules you shared had support for `on_click`. Is there a way in Py3status to figure out which of the 2+ players that would be clicked? Also, what if a click is run from `py3-cmd`?

lasers commented:
>If desired, I can make a PR implementing this config option for every module...

If you desire it, then go and make it. FYI, I haven't used py3status for a long while, but can try to help.

>... How would buttons be done? Is there a way in Py3status to figure out which of the 2+ players that would be clicked? Also, what if a click is run from py3-cmd?

I made many things and some unreleased code / unmerged PRs may contain this.
Note: My memory is hazy and is probably 94.44% correct.

IIRC, essentially, it's done by making a composite first with (`py3.safe_format)` then you update that composite with [`index`](https://py3status.readthedocs.io/en/latest/dev-guide/writing-modules/#outputs)  -- using [`py3.composite_update(existing_composite, new_index_dict)`](https://py3status.readthedocs.io/en/latest/dev-guide/the-py3-helper/#composite_updateitem-update_dict-softfalse).  When you clicked something, `on_click` event will be returned -- and your index should be found in there too -- your index should be unique string too to make things simple / identifiable. 

In main method, you should create unique usable `index` string such as `firefox-1235`. I picked pid num as an example since you could open multiple firefox instances).

Inside `on_click`, you'll break up that index string into usable configs. I think you may have to do things like assigning `firefox-1235/play` to play button composite... then split that index to firefox-1235 (instance id) and `if index == "play"` (clicked).

Get a loop of players to work first, then it's a minimal fix to add `index` string.

This should work in `py3-cmd` too...  see `py3-cmd click` as an example....  it would look something like `py3-cmd click --button 1 --index "firefox-1234/pause"` --  OK... You can look at `timer` module too... Since it contains `index` in the composite there.

jdholtz commented:
@lasers, I finished your requested changes. The colors and indices work for each player separately. Should I document how to select the index? The index is the player name, not the player instance because playerctl does not track two of the same player (at least for browsers, so I assume every player).

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,268 @@
+
+            format_player = self.py3.safe_format(self.format_player, player_data)
+            self.py3.composite_update(format_player, {"index": player_data["player"]})
+            self.py3.composite_update(format_player, {"color": color})
```
Remove in favor of using `\?color=status {title}` + thresholds.

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,268 @@
+            self.py3.composite_update(format_player, {"color": color})
+            players.append(format_player)
+
+        format_separator = self.py3.safe_format(self.format_separator)
```
(Silly) Please rename to `format_player_separator`.

If there is more than 1 `format_` loops (such as `format_player`, `format_song`, `format_tag`, whatever) in this module (or any other module), having a single separator config to control two or more separators can be unwanted... so separators for all `format_` loops ought to be kept separately too.

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,268 @@
+    def _on_name_appeared(self, manager, player_name):
+        self._init_player(player_name)
+
+    def _on_name_vanished(self, manager, name):
+        self.py3.update()
+
+    def _on_metadata(self, player, metadata, manager):
+        self.py3.update()
+
+    def _status_changed(self, player, status, manager):
+        self.py3.update()
```
About timing... Using playerctl command, it would make sense... With python bindings,  we shouldn't need `cache_timeout` and `sleep_timeout` config anymore. 

Fix to return  `{"cached_until": self.py3.CACHE_FOREVER, ...` instead.

jdholtz replied:
The refresh is there to update the position in the song/video (which is a valid attribute). Otherwise, it wouldn't update at all. What I can do is add a timeout only if the position is a placeholder and then set a timeout based on that (could be done in the `post_config_hook`).

lasers replied:
Is this confirmed? So far, it looks like this module is being updated consistently even with `self.py3.CACHE_FOREVER` tossed in there due to metadata/playback-status (possibly on 3 players).

jdholtz replied:
It updates whenever the metadata or playback status changes, but the `position` placeholder is also offered in this module. When I add that to the `format_player` and use `"cached_until": self.py3.CACHE_FOREVER`, the position does not update unless the metadata/status changes.

lasers replied:
That sounds so gooood. We might not need cache_timeout after all too. If position is included, just hardcode it to a second, then when players aren't tracked, cache_forever again.

lasers reviewed (changes requested):
Not tested yet.  Progress looks good.

lasers reviewed (commented):
See `git grep -A15 'class Meta' py3status/modules/loadavg.py` for amending `:escape` to specific placeholder(s).

jdholtz commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,268 @@
+        self.color_playing = self.py3.COLOR_PLAYING or self.py3.COLOR_GOOD
+        self.color_stopped = self.py3.COLOR_STOPPED or self.py3.COLOR_BAD
+
+        self._init_manager()
+
+    def _init_manager(self):
```
Do you mean just to remove the function and make the `_init_manager` code a part of the `post_config_hook` function? I split it up to make it more logically grouped, but looking at the` `_init_manager` function now, only a small portion actually deals with the manager.

lasers replied:
Yes. I thought that might not be needed (ie conky module just have everything in post_config_hook with meaningful comments) and it's not used anywhere else.

jdholtz commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,268 @@
+
+        return position
+
+    @staticmethod
+    def _microseconds_to_time(microseconds):
+        seconds = microseconds // 1_000_000
+        m, s = divmod(seconds, 60)
+        h, m = divmod(m, 60)
+        time = f"{h}:{m:02d}:{s:02d}"
+        return time.lstrip("0").lstrip(":")
```
I agree. It is currently used in `cmus`, `battery_level` (a tiny bit different but it probably should change to this format), and this module. Currently, `spotify` and `mpris` use Python's `datetime.timedelta` (which this function could actually use because it does the same thing. It just needs to use `lstrip` twice as well) to convert it into a valid time so they could change to this py3 function.

jdholtz commented:
I decided to automatically add `:escape` to title, based on your suggestions. Additionally, due to f-string formatting, I don't think the maximum length for a module is necessary because users have more precision over it using f-strings.

lasers commented:
>I decided to automatically add :escape to title

Nice on adding `:escape`.... Making it easier for users. :-) 

>Is there a reason re.match is not used instead?

No particular reason other than fnmatch is being used in many modules already (and possibly not needed). If you need to negate something, then we can roll back. Not familiar with playerctld. I assume users would do no more than to target their favorite players or just ignore that one player (ie wanting all information from playerctl).

> ...i3bar exiting unexpectedly... 

I see your issue. I wonder if it had to do something with Glib/Thread loop.
`git grep -Iri glib py3status/modules/*.py` shows slightly different ways to run thread loops. I wonder if it's the current  implementation that's causing the crash on exit... I don't recall crashes with conky module too. 

I toyed with your module some in virtual machine. For me, it is chromium that crashes it. Firefox does it okay. I don't think it's worth creating individual buttons (like mpris) since clicking full player does the job and is friendly enough.

This is going to be a nice module for many users... and we probably should remove `player_control` module too in favor of this one.... I don't even think that module works right now anyway. :-) 

I think we need to look at `cache_timeout / sleep_timeout` again too... I'm not seeing your description on how it works when I use `self.py3.CACHE_FOREVER` in there.

Imho, I would add `[{title}|{player}]` in your `format_player` to show player name if no title (ie vlc) too.

jdholtz commented:
I fixed the crash that was happening by adding a very small delay (`time.sleep(0.01)`) before updating the module after a player vanishes. This might not be the best way to do it, but it works perfectly for me and such a short delay would make no noticeable difference to the user.

I found out that the reason for this is because Playerctl is not given enough time to remove the player from the manager before this module retrieves tracked players. It would then error when getting `player.props.metadata` as the metadata no longer existed. 

Additionally, I removed the `_init_manager` function and added a default `{player}` to the `format_player` when no title exists.

For right now, I opted to remove the `sleep_timeout` when no players are tracked in favor of `self.py3.CACHE_FOREVER`. To optimize this even further, I could also check if the `{position}` exists in either of the placeholders and only set `cached_until` to `self.cache_timeout` if it does (if it doesn't exist, `CACHE_FOREVER` would be used).

jdholtz commented:
Hey @lasers, I believe I have addressed every one of your points. Please let me know if you have any other suggestions. Thanks very much for helping improve this module a lot!

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,272 @@
+
+        # Player attributes
+        data["player"] = player.props.player_name
+        data["status"] = player.props.status.lower()
```
(Change) Thoughts on sticking with default? Drop unnecessary `lower()` and update strings in `format_player` and thresholds. Make `{status}` usable too with `[{status}]` --> `[Playing]`, `[Paused]`, `[Stopped]` example.

It is what it is. True data. It just make `format_player` a bit more raw. Use it for a while and see how you feel about this change versus lowercase status.

jdholtz replied:
I'm good with doing that. It is already usable with `[{status}]` as it is a valid placeholder that can be used.

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,272 @@
+            ),
+        }
+
+    def _get_player_from_index(self, index):
+        for player in self.manager.props.players:
+            if player.props.player_name == index:
+                return player
+
+        return None
```
(Nitpick) Please move this func up. I care that last two funcs are for py3status's... and other `underscore_funcs` stays grouped on top. _....AKA I like to keep them separate._

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,272 @@
+        tracked_players = self.manager.props.players
+
+        # Only have a cache timeout if the user wants to know the players' positions
+        if tracked_players and (
+            self.py3.format_contains(self.format, "position")
+            or self.py3.format_contains(self.format_player, "position")
```
> self.py3.format_contains(self.format, "position")

Never can happen. No data. Remove.

> or self.py3.format_contains(self.format_player, "position"

Put this in `post_config_hook` and assign it to `self.position`. Bypass using `format_contains` on every update because the returned value would always be same. Do that once instead.

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,272 @@
+            # Playerctl doesn't support getting positions for all players
+            microseconds = player.get_position()
+            position = self._microseconds_to_time(microseconds)
+        except GLib.GError:
+            position = None
+
+        return position
```
(Comment) I thought for a brief second to make even smarter `cache_timeout` here... via using `any([pos1, pos2, pos3])`... but I think that's an overkill because 9 out of 10 decent players will have positions.

Talking about getting at least two players with no position. Probably won't last long anyway.

jdholtz replied:
What can be done is checking if any of the players are playing and have a valid position. This would be the only case I see that `cache_timeout` shouldn't be `CACHE_FOREVER`.

lasers replied:
It's your decision. Worth it or not. Idk.

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,272 @@
+SAMPLE OUTPUT
+[
+    {'color': '#00FF00', 'full_text': '> My Lucky Pants Failed Me Again - Tom Rosenthal'}
+    {'color': '#FFFF00', 'full_text': '|| Too Much Skunk Tonight - Birdy Nam Nam'}
```
One more line for stopped, plz.

lasers reviewed (commented):
Did this get everything it can? (shuffle, loop, volume status, mute, et cetera).

Scrolling mouse on player to seek forward / backward?
Scrolling mouse to change volume?
Clicking to mute? Things like that.

Remember, we want to be better than mpris / player_control if we can.  ;-) 

Otherwise, it looks great to go after one or two more iterations.

jdholtz commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,272 @@
+            self.py3.format_contains(self.format, "position")
+            or self.py3.format_contains(self.format_player, "position")
+        ):
+            cached_until = 1
```
Sounds good. Is there a reason `cache_timeout` can't just be declared at the top of the class by the button and format default configurations instead of using `getattr`?

lasers replied:
Because once you expose a config, it can be hard to carry a config over to new one. This is more for `format_` related configs...

I think it would good to give users an impression that playerctl module's updates are decided automatically based on data changes/actions rather than at interval. If we expose this, then users may think this is done at interval.

jdholtz commented:
Hey @lasers, I finished adding support for everything I think is possible with Playerctl. This includes displaying and changing the volume, shuffle mode, loop status, and seeking forward/backward with buttons.

While doing this, I noticed a few odd behaviors. Cmus changes the volume and loop status (not shuffle) using playerctl, but playerctl does not respond to updates from the player, so it effectively does not work. For example, to increase the volume, the current volume is added to a `volume_change` configuration value. However, the volume for the player will always be the same so it will only change once. This seems to only be an issue with Cmus though as Spotify works perfectly for all of these new controls.

I also looked into how to mute the players, but there is no specific functionality to mute using MPRIS. Setting the volume to 0 doesn't work because this module doesn't remember the previous volume (cmus also doesn't mute correctly natively). If you have any suggestions to do this, I would be happy to add them. However, it seems infeasible because it depends on the player's volume integrating well with playerctl and the volume before mute might need to persist even when py3status restarts.

Last, I added a better caching algorithm so the module only updates when necessary.

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -65,22 +79,30 @@ class Py3status:
     thresholds = {
         "status": [("Playing", "good"), ("Paused", "degraded"), ("Stopped", "bad")]
     }
+    volume_change = 10
```
(Suggestion)
Rename to:
- `button_shuffle`
- `seek_delta`
- `volume_delta`
- `button_loop_status` Current naming looks fine following playerctl func too. This is me trying to keep button config short / simple... and to rearrange button configs too.

Honestly, it looks absolutely ready enough. Nice looping on checking position.

I can't remember, but I think the problem with cmus was that it's still using mpris1.

jdholtz replied:
What's strange with cmus is that it works when I use the playerctl CLI, but it doesn't when I use the library (the CLI runs the same function as the library though). I filed a bug report on playerctl for this inconsistent behavior, but there hasn't been any activity on the project in almost two years so it might never be looked at.

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,360 @@
+        format_player_separator = self.py3.safe_format(self.format_player_separator)
+        format_players = self.py3.composite_join(format_player_separator, players)
+
+        # Only set a cache timeout if necessary
+        if not cache_forever and self.position:
+            cached_until = self.cache_timeout
+        else:
+            cached_until = self.py3.CACHE_FOREVER
```
I see this , I see two different conditionals for `cached_*`.

Instead of using slightly redundant `cache_forever = True` above, can you use `cached_until = self.py3.CACHE_FOREVER` instead -- and use `if self.position and .... etc` in the loop?

jdholtz commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,358 @@
+        time.sleep(0.01)
+        self.py3.update()
+
+    def _loop_status_changed(self, player, loop_status, manager):
```
I could point these event handlers (loop, volume, metadata, etc.) to the same function to simplify the code a little. However, that might not be best practice since the parameters are of different types (even though it is the same number of them). Additionally, it wouldn't allow extensibility in the future.

lasers reviewed (approved):
I guess it couldn't be helped with excessive buttons...  Sometimes I wonder having excessive buttons in the modules is one of annoying thing about modules... and thought about making `buttons = {"volume_up": 4, "volume_down": 5}` dict config to make it go from from 12 down to 1. I think that could be not friendly for users too.

Everything playerctl offer is now possibly supported by this module too.

Nice job. 🥇

lasers commented:
You should make a PR to remove `player_control` module too.

jdholtz commented:
I wouldn't be opposed to adding a dictionary for the buttons. However, it doesn't follow the configuration of other modules so it might be more confusing to users than just having 12 button configuration values.

jdholtz commented:
> You should make a PR to remove player_control module too. 

There are a few other modules that could possibly be removed as well. 
1. `gpmdp` -- playerctl supports Google Play Music Desktop Player (AFAIK from the repository issues)
2. `mpris` -- this is basically the `playerctl` module using DBus instead of Playerctl (but with less functionality)
3. `clementine` -- playerctl supports the Clementine music player
4. `spotify` -- uses a DBus implementation to access Spotify

All of the above don't appear to provide any more data/functionality that this module can provide. I don't think `cmus` should be removed because it provides a lot more information than Playerctl can get.

Is there a process to removing modules, similar to deprecating configs?

lasers commented:
Not linking to unrelated pulls.
```
nvidia_temp -- https://github.com/ultrabug/py3status/pull/1853 
bitcoin_price -- https://github.com/ultrabug/py3status/pull/2047
...etc cetera
```

ultrabug commented:
Hello @jdholtz ; thanks for this amazing work and first contribution to py3status (thanks to you too @lasers)

I had some work to do on this project which I hope has been covered yesterday with the 3.52 release.

I'll do my best to look at and test your module quickly.

Kindly

ultrabug commented:
Meanwhile, I would be grateful if you could rebase your branch with current master (I changed quite a lot of things regarding to CI and build system).

jdholtz commented:
Thanks @ultrabug. I rebased my branch, so let me know if anything else should be done.

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,358 @@
+        }
+
+    def post_config_hook(self):
+        self.color_paused = self.py3.COLOR_PAUSED or self.py3.COLOR_DEGRADED
+        self.color_playing = self.py3.COLOR_PLAYING or self.py3.COLOR_GOOD
+        self.color_stopped = self.py3.COLOR_STOPPED or self.py3.COLOR_BAD
+
```
I think we don't need them anymore with thresholds in place.

jdholtz commented:
This is more of a configuration question (I didn't want to open a new issue for it. Maybe opening GitHub discussions could help users with similar questions), but how can I use multiple format specifiers? I want to escape the title and limit it to 15 characters, but my current configuration does not work even though it does in Python's f-string formatting (which this appears to be based off of). Here is what I tried: `{title:escape.15s}`. `{title:.15s}` works, but then it gets rid of the default escape. Is this possible yet in Py3Status?

lasers commented:
IIRC it's not from Python f-string formatting. Several things were coded in to behave like that.

As for length, you may want `max_length`.... See `git grep _length ~/src/py3status/tests` -- so length is not limited to placeholders only, but the whole block if we add things inside the brackets.

jdholtz commented:
Wrapping the title in a max_length block worked. Thanks!

lasers commented on py3status/modules/playerctl.py:
```diff
@@ -0,0 +1,354 @@
+        self._set_data_from_metadata(player, data)
+
+        # Player attributes
+        data["player"] = player.props.player_name
+        data["loop_status"] = player.props.loop_status.value_nick
```
I see it's actually `player_name` being trimmed. Thoughts on doing it same for `button_loop` and `{loop}`?

jdholtz commented:
@ultrabug any updates to this? I’ve been using it for over a month now and it has worked very well for me.

ultrabug commented:
Thanks a lot @jdholtz and @lasers 

I'm merging tho CI fails, I'll fix that myself before releasing a new version of py3status

jdholtz commented:
Thanks @ultrabug. The code QA commit did wrongly move an import from line 75 to 73 in [this commit](https://github.com/ultrabug/py3status/commit/3bc4946f6c60f066389ab72a62cc56e2d079f3ed). When running the module in test mode (`python playerctl.py`), it prints the message `/home/jdholtz/.config/py3status/modules/playerctl.py:73: PyGIWarning: Playerctl was imported without specifying a version first. Use gi.require_version('Playerctl', '2.0') before import to ensure that the right version gets loaded.` which is why I had the import after. However, it isn't a big deal since it cannot be seen in the log file and doesn't cause any issues.

Also, let me know if I should submit any PRs to deprecate modules that are similar to Playerctl's functionality but are old. I mentioned possible module deprecations [here](https://github.com/ultrabug/py3status/pull/2211#issuecomment-1654994490).

ultrabug commented:
@jdholtz indeed, CI was not happy about it. So if that doesn't break anything let's leave it.

> Also, let me know if I should submit any PRs to deprecate modules that are similar to Playerctl's functionality but are old. I mentioned possible module deprecations here.

It's hard to know which modules are used or not. I guess clementine and gpmdp are very old and unmaintained.

I'm sure that spotify and mpris have a strong user base so I would not touch them.
